File: graph/undirected-path/undirected-path.js
// Write a function, undirectedPath, that takes in an array of edges for an undirected graph and two nodes (nodeA, nodeB). The function should return a boolean indicating whether or not there exists a path between nodeA and nodeB.


const undirectedPath = (edges, nodeA, nodeB) => {
    const graph = edgesToGraph(edges);
    const newSet = new Set();
    const stack = [nodeA];

    while (stack.length > 0) {
        const cur = stack.pop();
        if (newSet.has(cur)) {
            continue;
        }
        if (cur === nodeB) {
            return true;
        }
        newSet.add(cur);
        for (let node of graph[cur]) {
            stack.push(node);
        }
    }
    return false;
};

const edgesToGraph = edges => {
    const graph = {};
    for (let edge of edges) {
        for (let i = 0; i < edge.length; i++) {
            if (graph[edge[i]]) {
                if (i === 0) {
                    graph[edge[i]].push(edge[1]);
                } else {
                    graph[edge[i]].push(edge[0]);
                }
            } else {
                if (i === 0) {
                    graph[edge[i]] = [edge[1]];
                } else {
                    graph[edge[i]] = [edge[0]];
                }
            }
        }
    }
    return graph;
};


const edges = [
    ['b', 'a'],
    ['c', 'a'],
    ['b', 'c'],
    ['q', 'r'],
    ['q', 's'],
    ['q', 'u'],
    ['q', 't'],
];

console.log(undirectedPath(edges, 'r', 'b'));
// -> true

File: 4.graph/3.undirected-path/undirected-path.go
// Write a function, undirectedPath, that takes in an array of edges for an undirected graph and two nodes (nodeA, nodeB). The function should return a boolean indicating whether or not there exists a path between nodeA and nodeB.

package main

import "fmt"

type Graph struct {
	Nodes map[string][]string
}

func undirectedPath(edges [][]string, nodeA, nodeB string) bool {
	graph := edgesToGraph(edges)
	newSet := make(map[string]struct{})
	stack := []string{nodeA}
	for len(stack) > 0 {
		cur := stack[len(stack)-1]
		if _, ok := newSet[cur]; ok {
			stack = stack[:len(stack)-1]
			continue
		}
		if cur == nodeB {
			return true
		}
		newSet[cur] = struct{}{}
		stack = stack[:len(stack)-1]
		for _, node := range (*graph).Nodes[cur] {
			stack = append(stack, node)
		}
	}
	return false
}

func edgesToGraph(edges [][]string) *Graph {
	graph := &Graph{
		Nodes: make(map[string][]string),
	}
	for _, edge := range edges {
		for i, node := range edge {
			nodes := graph.Nodes[node]
			if i == 0 {
				nodes = append(nodes, edge[1])
			} else {
				nodes = append(nodes, edge[0])
			}
			graph.Nodes[node] = nodes
		}
	}
	return graph
}

func main() {
	edges := [][]string{
		{"i", "j"},
		{"k", "i"},
		{"m", "k"},
		{"k", "l"},
		{"o", "n"},
	}

	fmt.Println(undirectedPath(edges, "j", "m"))
}
